Engine/gfx/ali3dogl.cpp

```diff
@@ -1397,166 +1396,40 @@ void OGLGraphicsDriver::DestroyDDB(IDriverDependantBitmap* bitmap)
   delete bitmap;
 }
 
-__inline void get_pixel_if_not_transparent15(unsigned short *pixel, unsigned short *red, unsigned short *green, unsigned short *blue, unsigned short *divisor)
-{
-  if (pixel[0] != MASK_COLOR_15)
-  {
-    *red += algetr15(pixel[0]);
-    *green += algetg15(pixel[0]);
-    *blue += algetb15(pixel[0]);
-    divisor[0]++;
-  }
-}
 
-__inline void get_pixel_if_not_transparent32(unsigned int *pixel, unsigned int *red, unsigned int *green, unsigned int *blue, unsigned int *divisor)
-{
-  if (pixel[0] != MASK_COLOR_32)
-  {
-    *red += algetr32(pixel[0]);
-    *green += algetg32(pixel[0]);
-    *blue += algetb32(pixel[0]);
-    divisor[0]++;
-  }
-}
-
-#define D3DCOLOR_RGBA(r,g,b,a) \
-  (((((a)&0xff)<<24)|(((b)&0xff)<<16)|(((g)&0xff)<<8)|((r)&0xff)))
-
-
-void OGLGraphicsDriver::UpdateTextureRegion(TextureTile *tile, Bitmap *bitmap, OGLBitmap *target, bool hasAlpha)
+void OGLGraphicsDriver::UpdateTextureRegion(OGLTextureTile *tile, Bitmap *bitmap, OGLBitmap *target, bool hasAlpha)
 {
   int textureHeight = tile->height;
   int textureWidth = tile->width;
 
+  // TODO: this seem to be tad overcomplicated, these conversions were made
+  // when texture is just created. Check later if this operation here may be removed.
   AdjustSizeToNearestSupportedByCard(&textureWidth, &textureHeight);
 
   int tileWidth = (textureWidth > tile->width) ? tile->width + 1 : tile->width;
   int tileHeight = (textureHeight > tile->height) ? tile->height + 1 : tile->height;
 
   bool usingLinearFiltering = _filter->UseLinearFiltering();
   bool lastPixelWasTransparent = false;
-  char *origPtr = (char*)malloc(4 * tileWidth * tileHeight);
+  char *origPtr = (char*)malloc(sizeof(int) * tileWidth * tileHeight);
   char *memPtr = origPtr;
-  for (int y = 0; y < tileHeight; y++)
+
+  TextureTile fixedTile;
+  fixedTile.x = tile->x;
+  fixedTile.y = tile->y;
+  fixedTile.width = Math::Min(tile->width, tileWidth);
+  fixedTile.height = Math::Min(tile->height, tileHeight);
+  int pitch = tileWidth * sizeof(int);
+  BitmapToVideoMem(bitmap, hasAlpha, &fixedTile, target, memPtr, pitch, usingLinearFiltering);
+
+  // Mimic the behaviour of GL_CLAMP_EDGE for the bottom line
+  if (tile->height < tileHeight)
   {
-    // Mimic the behaviour of GL_CLAMP_EDGE for the bottom line
-    if (y == tile->height)
-    {
-      unsigned int* memPtrLong = (unsigned int*)memPtr;
-      unsigned int* memPtrLong_previous = (unsigned int*)(memPtr - tileWidth * 4);
+    unsigned int* memPtrLong = (unsigned int*)(memPtr + pitch * tile->height);
+    unsigned int* memPtrLong_previous = (unsigned int*)(memPtr + pitch * (tile->height - 1));
 
       for (int x = 0; x < tileWidth; x++)
         memPtrLong[x] = memPtrLong_previous[x] & 0x00FFFFFF;
-
```

Comment:
Is it possible to revert some of these refactorings until later? I believe some of these changes that are simply refactoring/tidying up will conflict with the sdl2 port.

Reply:
That's possible, but what's the nature of conflict? I looked into the latest code in your sdl2 branch and it seem identical to what is in 3.4.1 now.